src/server.c

```diff
@@ -1169,16 +1169,43 @@ void cronUpdateMemoryStats() {
 /* Pause clients on write as long as OOM and pending lazy-free jobs in background.
  * By doing so the server might avoid unnecessary eviction or rejection. */
 static void clientWritePauseDuringOOM() {
+    int is_paused = (getPauseActionsPurposeEndTime(PAUSE_OOM_THROTTLE) == LLONG_MAX);
+    enum { ACTION_NONE, 
+           ACTION_START_PAUSE,
+           ACTION_END_PAUSE
+    } action = ACTION_NONE;
```

Comment:
do we really need all that fancy enum and goto?

can't we just add something defrag.c does the bottom (instead of return):
```c
is_paused = getPauseActionsPurposeEndTime();
if (is_paused) {
    if (!stat_current_client_paused_oom_time)
        elapsedStart(&stat_current_client_paused_oom_time)
} else if (stat_current_client_paused_oom_time) {
    stat_total_client_paused_oom_time += elapsedUs(stat_current_client_paused_oom_time);
    server.stat_current_client_paused_oom_time = 0;
}
```

i think it's simpler since that piece of code is independent and doesn't rely on any hints from above.
i don't see any major issues with RESETSTAT (unlike your code, after reset, if pause is in effect, it'll start measuring immediately and not on next pause).
and the bonus is that it's similar to the other mechanisms, so it's behavior is less likely to surprise anyone.

Reply:
Cool. Please note that your code avoid early return in favor of updating statistics soon after RESETSTAT.